src/workers/filesystems/infrastructure/SyncRemoteFileSystem.ts

```diff
@@ -0,0 +1,666 @@
+              `oldName: ${oldName}, newName: ${newName}, fileInCache: ${JSON.stringify(
+                fileInCache,
+                null,
+                2
+              )}, res: ${await serializeRes(res)}`
+            )
+          );
+        }
+        delete cache[oldName];
+        cache[newName] = fileInCache;
+      } catch (err) {
+        await handleFetchError(
+          err,
+          'Renaming remote file',
+          `oldName: ${oldName}, newName: ${newName}, fileInCache: ${JSON.stringify(
+            fileInCache,
+            null,
+            2
+          )}`
+        );
+      }
+    },
+
+    async pullFile(
+      name: string,
+      source: Source,
+      progressCallback: (progress: number) => void
+    ): Promise<void> {
+      const { size, modTime: modTimeInSeconds } = source;
+      const route = name.split('/');
+
+      const { name: baseNameWithoutExt, ext } = path.parse(
+        route.pop() as string
+      );
+      const fileType = ext.slice(1);
+
+      let lastParentId = baseFolderId;
+
+      if (route.length > 0) {
+        for (const [i, folderName] of route.entries()) {
+          const routeToThisPoint = route.slice(0, i + 1).join('/');
+
+          const folderInCache = cache[routeToThisPoint];
+
+          if (folderInCache) lastParentId = folderInCache.id;
+          else {
+            if (createFolderQueue.listeners(routeToThisPoint).length === 0) {
+              httpRequest(`${process.env.API_URL}/api/storage/folder`, {
+                method: 'POST',
+                headers,
+                body: JSON.stringify({
+                  folderName,
+                  parentFolderId: lastParentId,
+                }),
+              })
+                .then(async (res) => {
+                  if (!res.ok) {
+                    throw new ProcessError(
+                      'BAD_RESPONSE',
+                      createErrorDetails(
+                        {},
+                        'Creating folder in drive server',
+                        `res: ${await serializeRes(
+                          res
+                        )}, folderName: ${folderName}, parentFolderId: ${lastParentId}`
+                      )
+                    );
+                  } else return res;
+                })
+                .then((res) => res.json())
+                .then((createdFolder: ServerFolder) => {
+                  cache[routeToThisPoint] = {
+                    id: createdFolder.id,
+                    parentId: createdFolder.parent_id as number,
+                    isFolder: true,
+                    bucket: createdFolder.bucket,
+                  };
+                  createFolderQueue.emit(routeToThisPoint, createdFolder.id);
+                })
+                .catch(async (err) => {
+                  createFolderQueue.emit(routeToThisPoint, err);
+                });
+            }
+
+            try {
+              lastParentId = await new Promise((resolve, reject) =>
```

Comment:
I see what are you doing here. You are not using async and await to have time to listen to the event finished. This approach could be done, but **always** put the listener **before** the emit that you are doing, otherwise the risk of missing an event is unavoidable. 

Suggestion: I would recommend using `queue` from the `async` library which is a queue that works with async-await or callbacks and can do what you are trying to do here in a cleaner, abstracted and easier-to-use way.